assets/js/upload.js

```diff
@@ -171,9 +171,127 @@ function setupImageUpload() {
     window.removeEventListener('beforeunload', beforeUnload);
   }
 
+  function createTagError(message) {
+    const buttonAfter = $('#tagsinput-save');
+    const errorElement = makeEl('span', { className: 'help-block tag-error' });
+
+    errorElement.innerText = message;
+    buttonAfter.parentElement.insertBefore(errorElement, buttonAfter);
+  }
+
+  function clearTagErrors() {
+    const tagErrorElements = $$('.tag-error');
+
+    // remove() causes the length to decrease
+    while (tagErrorElements.length > 0) {
+      tagErrorElements[0].remove();
+    }
+  }
+
+  // populate tag error helper bars as necessary
+  // return true if all checks pass
+  // return false if any check fails
+  function validateTags() {
+    const tags = $$('.tag');
+
+    if (tags.length === 0) {
+      createTagError('Tag input must contain at least 3 tags');
+      return false;
+    }
+
+    const tagsArr = [];
+
+    for (const i in tags) {
+      let tag = tags[i].innerText;
+
+      tag = tag.substring(0, tag.length - 2); // remove " x" from the end
+      tagsArr.push(tag);
+    }
+
+    const ratingsTags = ['safe', 'suggestive', 'questionable', 'explicit', 'semi-grimdark', 'grimdark', 'grotesque'];
+
+    const errors = [];
+
+    let hasRating = false;
+    let hasSafe = false;
+    let hasOtherRating = false;
+
+    tagsArr.forEach(tag => {
+      if (ratingsTags.includes(tag)) {
+        hasRating = true;
+        if (tag === 'safe') {
+          hasSafe = true;
+        } else {
+          hasOtherRating = true;
+        }
+      }
+    });
+
+    if (!hasRating) {
+      errors.push('Tag input must contain at least one rating tag');
+    } else if (hasSafe && hasOtherRating) {
+      errors.push('Tag input may not contain any other rating if safe');
+    }
+
+    if (tagsArr.length < 3) {
+      errors.push('Tag input must contain at least 3 tags');
+    }
+
+    errors.forEach(msg => createTagError(msg));
+
+    return errors.length === 0; // true: valid if no errors
+  }
+
+  function enableUploadButton() {
+    const submitButton = $('.input--separate-top');
+    if (submitButton !== null) {
+      submitButton.disabled = false;
+      submitButton.innerText = 'Upload';
+    }
+  }
+
+  function disableUploadButton() {
+    const submitButton = $('.input--separate-top');
+    if (submitButton !== null) {
+      submitButton.disabled = true;
+      submitButton.innerText = 'Please wait...';
+    }
+
+    // delay is needed because Safari stops the submit if the button is immediately disabled
+    requestAnimationFrame(() => submitButton.setAttribute('disabled', 'disabled'));
+  }
+
+  function anchorToTop() {
+    let url = window.location.href;
+    url = url.split('#')[0]; //remove any existing hash anchor from url
+    url += '#'; //move view to top of page
+    window.location.href = url;
+  }
```

Comment:
Perhaps use `window.scrollTo({ top: 0, left: 0 });` instead of modifying the anchor?

Reply:
I really wanted to anchor to a tags element, that way users could immediately see the error and have a context cue

Reply:
You could also use .scrollIntoView

Reply:
And if none of those feel right to you, it is possible to [explicitly set window.location.hash](https://html.spec.whatwg.org/multipage/nav-history-apis.html#dom-location-hash-dev)

Reply:
now using .scrollIntoView with a fallback to window.scrollTo top left